rasa/cdu/commands/correct_slots_command.py

```diff
@@ -79,13 +47,190 @@ def from_dict(cls, data: Dict[str, Any]) -> CorrectSlotsCommand:
         Returns:
             The converted dictionary.
         """
-        return CorrectSlotsCommand(
-            corrected_slots=[
-                CorrectedSlot(s["name"], value=s["value"])
-                for s in data["corrected_slots"]
-            ]
+        try:
+            return CorrectSlotsCommand(
+                corrected_slots=[
+                    CorrectedSlot(s["name"], value=s["value"])
+                    for s in data["corrected_slots"]
+                ]
+            )
+        except KeyError as e:
+            raise ValueError(
+                f"Missing key when parsing CorrectSlotsCommand: {e}"
+            ) from e
+
+    @staticmethod
+    def are_all_slots_reset_only(
+        proposed_slots: Dict[str, Any], all_flows: FlowsList
+    ) -> bool:
+        """Checks if all slots are reset only.
+
+        A slot is reset only if the `collect_information` step it gets filled by
+        has the `ask_before_filling` flag set to `True`. This means, the slot
+        shouldn't be filled if the question isn't asked.
+
+        If such a slot gets corrected, we don't want to correct the slot but
+        instead reset the flow to the question where the slot was asked.
+
+        Args:
+            proposed_slots: The proposed slots.
+            all_flows: All flows in the assistant.
+
+        Returns:
+            `True` if all slots are reset only, `False` otherwise.
+        """
+        return all(
+            collect_information_step.collect_information not in proposed_slots
+            or collect_information_step.ask_before_filling
+            for flow in all_flows.underlying_flows
+            for collect_information_step in flow.get_collect_information_steps()
         )
 
+    @staticmethod
+    def find_earliest_updated_collect_info(
+        user_frame: UserFlowStackFrame,
+        updated_slots: List[str],
+        all_flows: FlowsList,
+    ) -> Optional[FlowStep]:
+        """Find the earliest collect information step that fills one of the slots.
+
+        When we update slots, we need to reset a flow to the question when the slot
+        was asked. This function finds the earliest collect information step that
+        fills one of the slots - with the idea being that we afterwards go through
+        the other updated slots.
+
+        Args:
+            user_frame: The current user flow frame.
+            updated_slots: The slots that were updated.
+            all_flows: All flows.
+
+        Returns:
+        The earliest collect information step that fills one of the slots.
+        """
+        flow = user_frame.flow(all_flows)
+        step = user_frame.step(all_flows)
+        asked_collect_info_steps = flow.previous_collect_information_steps(step.id)
```

Comment:
This goes beyond this PR... I am wondering whether this approximate way of creating previous collect information steps will result in any issues. Currently we move backwards from a given step id but this could technically result in wrong results in cases of branches merging again. If you call this method on a merge step or after it, you'll get all slots from both branches, although there was only one taken. You could get this in our verify account flow if you call this on the final confirm step. I think you'll get the income question even if you went the not based in ca route. Maybe it's not a problem. The way to get the exact set of slots would probably simulate the flow forwards from the starting step. Given the current slots you could chart the path to the current step id.

Reply:
yes totally agreed. I think we also discussed that when we discussed the jumping back behaviour. 

I can add a todo for now to mark it as something that might need another thought.